File: lib/src/features/member_staff/widgets/staff_rating_summary.dart
import 'package:flutter/material.dart';
import 'package:flutter_rating_bar/flutter_rating_bar.dart';
import '../models/rating_summary.dart';
import '../providers/staff_rating_provider.dart';
import 'package:provider/provider.dart';

class StaffRatingSummary extends StatefulWidget {
  final int staffId;
  final String staffType;
  final VoidCallback? onRatePressed;

  const StaffRatingSummary({
    Key? key,
    required this.staffId,
    required this.staffType,
    this.onRatePressed,
  }) : super(key: key);

  @override
  _StaffRatingSummaryState createState() => _StaffRatingSummaryState();
}

class _StaffRatingSummaryState extends State<StaffRatingSummary> {
  late Future<RatingSummary> _summaryFuture;

  @override
  void initState() {
    super.initState();
    _loadRatingSummary();
  }

  void _loadRatingSummary() {
    final ratingProvider = Provider.of<StaffRatingProvider>(context, listen: false);
    _summaryFuture = ratingProvider.getRatingSummary(
      staffId: widget.staffId,
      staffType: widget.staffType,
    );
  }

  @override
  Widget build(BuildContext context) {
    return FutureBuilder<RatingSummary>(
      future: _summaryFuture,
      builder: (context, snapshot) {
        if (snapshot.connectionState == ConnectionState.waiting) {
          return _buildLoadingState();
        } else if (snapshot.hasError) {
          return _buildErrorState(snapshot.error.toString());
        } else if (snapshot.hasData) {
          return _buildSummaryCard(snapshot.data!);
        } else {
          return _buildEmptyState();
        }
      },
    );
  }

  Widget _buildLoadingState() {
    return Card(
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Center(
          child: CircularProgressIndicator(),
        ),
      ),
    );
  }

  Widget _buildErrorState(String error) {
    return Card(
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(Icons.error_outline, color: Colors.red, size: 48),
            SizedBox(height: 8),
            Text(
              'Failed to load ratings',
              style: TextStyle(fontWeight: FontWeight.bold),
            ),
            SizedBox(height: 4),
            Text(
              error,
              style: TextStyle(fontSize: 12, color: Colors.grey[600]),
              textAlign: TextAlign.center,
            ),
            SizedBox(height: 8),
            ElevatedButton(
              onPressed: () {
                setState(() {
                  _loadRatingSummary();
                });
              },
              child: Text('Retry'),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildEmptyState() {
    return Card(
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            Icon(Icons.star_border, color: Colors.amber, size: 48),
            SizedBox(height: 8),
            Text(
              'No ratings yet',
              style: TextStyle(fontWeight: FontWeight.bold),
            ),
            SizedBox(height: 8),
            ElevatedButton(
              onPressed: widget.onRatePressed,
              child: Text('Be the first to rate'),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildSummaryCard(RatingSummary summary) {
    return Card(
      elevation: 2,
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      'Rating & Reviews',
                      style: TextStyle(
                        fontSize: 18,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    SizedBox(height: 4),
                    Text(
                      '${summary.totalRatings} ${summary.totalRatings == 1 ? 'review' : 'reviews'}',
                      style: TextStyle(
                        color: Colors.grey[600],
                        fontSize: 14,
                      ),
                    ),
                  ],
                ),
                ElevatedButton(
                  onPressed: widget.onRatePressed,
                  child: Text('Rate'),
                ),
              ],
            ),
            SizedBox(height: 16),
            Row(
              children: [
                Text(
                  summary.averageRating.toStringAsFixed(1),
                  style: TextStyle(
                    fontSize: 48,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                SizedBox(width: 16),
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    RatingBarIndicator(
                      rating: summary.averageRating,
                      itemBuilder: (context, index) => Icon(
                        Icons.star,
                        color: Colors.amber,
                      ),
                      itemCount: 5,
                      itemSize: 24.0,
                    ),
                    SizedBox(height: 4),
                    Text(
                      'Average rating',
                      style: TextStyle(
                        color: Colors.grey[600],
                        fontSize: 14,
                      ),
                    ),
                  ],
                ),
              ],
            ),
            SizedBox(height: 16),
            _buildRatingDistribution(summary.ratingDistribution),
            if (summary.recentReviews.isNotEmpty) ...[
              SizedBox(height: 16),
              Text(
                'Recent Reviews',
                style: TextStyle(
                  fontSize: 16,
                  fontWeight: FontWeight.bold,
                ),
              ),
              SizedBox(height: 8),
              ...summary.recentReviews.map((review) => _buildReviewItem(review)),
            ],
          ],
        ),
      ),
    );
  }

  Widget _buildRatingDistribution(Map<int, int> distribution) {
    final totalRatings = distribution.values.fold<int>(0, (sum, count) => sum + count);
    if (totalRatings == 0) return SizedBox.shrink();

    return Column(
      children: List.generate(5, (index) {
        final rating = 5 - index;
        final count = distribution[rating] ?? 0;
        final percentage = totalRatings > 0 ? count / totalRatings : 0.0;

        return Padding(
          padding: const EdgeInsets.symmetric(vertical: 2.0),
          child: Row(
            children: [
              Text(
                '$rating',
                style: TextStyle(
                  fontWeight: FontWeight.bold,
                  fontSize: 14,
                ),
              ),
              SizedBox(width: 8),
              Expanded(
                child: LinearProgressIndicator(
                  value: percentage,
                  backgroundColor: Colors.grey[200],
                  valueColor: AlwaysStoppedAnimation<Color>(Colors.amber),
                ),
              ),
              SizedBox(width: 8),
              Text(
                '$count',
                style: TextStyle(
                  color: Colors.grey[600],
                  fontSize: 14,
                ),
              ),
            ],
          ),
        );
      }),
    );
  }

  Widget _buildReviewItem(RecentReview review) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              CircleAvatar(
                child: Text(
                  review.memberName.substring(0, 1).toUpperCase(),
                ),
                radius: 16,
              ),
              SizedBox(width: 8),
              Text(
                review.memberName,
                style: TextStyle(
                  fontWeight: FontWeight.bold,
                ),
              ),
              Spacer(),
              Text(
                _formatDate(review.createdAt),
                style: TextStyle(
                  color: Colors.grey[600],
                  fontSize: 12,
                ),
              ),
            ],
          ),
          SizedBox(height: 4),
          RatingBarIndicator(
            rating: review.rating.toDouble(),
            itemBuilder: (context, index) => Icon(
              Icons.star,
              color: Colors.amber,
            ),
            itemCount: 5,
            itemSize: 16.0,
          ),
          if (review.feedback != null && review.feedback!.isNotEmpty) ...[
            SizedBox(height: 4),
            Text(
              review.feedback!,
              style: TextStyle(
                fontSize: 14,
              ),
            ),
          ],
          Divider(),
        ],
      ),
    );
  }

  String _formatDate(DateTime date) {
    final now = DateTime.now();
    final difference = now.difference(date);

    if (difference.inDays < 1) {
      return 'Today';
    } else if (difference.inDays < 2) {
      return 'Yesterday';
    } else if (difference.inDays < 7) {
      return '${difference.inDays} days ago';
    } else if (difference.inDays < 30) {
      final weeks = (difference.inDays / 7).floor();
      return '$weeks ${weeks == 1 ? 'week' : 'weeks'} ago';
    } else if (difference.inDays < 365) {
      final months = (difference.inDays / 30).floor();
      return '$months ${months == 1 ? 'month' : 'months'} ago';
    } else {
      final years = (difference.inDays / 365).floor();
      return '$years ${years == 1 ? 'year' : 'years'} ago';
    }
  }
}

File: admin-dashboard/components/StaffRatingSummary.js
import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { Rating } from '@mui/material';

export default function StaffRatingSummary({ staffId, staffType }) {
  const [summary, setSummary] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    fetchRatingSummary();
  }, [staffId, staffType]);

  const fetchRatingSummary = async () => {
    try {
      setLoading(true);
      const response = await axios.get(`/api/staff/${staffId}/ratings`, {
        params: { staff_type: staffType },
        headers: {
          Authorization: `Bearer ${localStorage.getItem('admin_token')}`
        }
      });
      
      setSummary(response.data);
      setError(null);
    } catch (error) {
      console.error('Error fetching rating summary:', error);
      setError('Failed to load rating summary');
    } finally {
      setLoading(false);
    }
  };

  if (loading) {
    return (
      <div className="bg-white p-4 rounded-lg shadow">
        <div className="animate-pulse flex space-x-4">
          <div className="flex-1 space-y-4 py-1">
            <div className="h-4 bg-gray-200 rounded w-3/4"></div>
            <div className="space-y-2">
              <div className="h-4 bg-gray-200 rounded"></div>
              <div className="h-4 bg-gray-200 rounded w-5/6"></div>
            </div>
          </div>
        </div>
      </div>
    );
  }

  if (error) {
    return (
      <div className="bg-white p-4 rounded-lg shadow">
        <div className="text-red-500">{error}</div>
      </div>
    );
  }

  if (!summary || summary.total_ratings === 0) {
    return (
      <div className="bg-white p-4 rounded-lg shadow">
        <h3 className="text-lg font-medium text-gray-900 mb-2">Ratings & Reviews</h3>
        <p className="text-gray-500">No ratings yet</p>
      </div>
    );
  }

  return (
    <div className="bg-white p-4 rounded-lg shadow">
      <h3 className="text-lg font-medium text-gray-900 mb-2">Ratings & Reviews</h3>
      
      <div className="flex items-center mb-4">
        <div className="text-4xl font-bold mr-4">{summary.average_rating}</div>
        <div>
          <Rating value={summary.average_rating} readOnly precision={0.1} />
          <div className="text-sm text-gray-500">{summary.total_ratings} {summary.total_ratings === 1 ? 'review' : 'reviews'}</div>
        </div>
      </div>
      
      {/* Rating Distribution */}
      <div className="mb-4">
        {[5, 4, 3, 2, 1].map(rating => {
          const count = summary.rating_distribution[rating] || 0;
          const percentage = summary.total_ratings > 0 
            ? (count / summary.total_ratings) * 100 
            : 0;
            
          return (
            <div key={rating} className="flex items-center mb-1">
              <div className="w-8 text-sm text-gray-600">{rating}</div>
              <div className="w-full bg-gray-200 rounded-full h-2.5 mr-2">
                <div 
                  className="bg-yellow-400 h-2.5 rounded-full" 
                  style={{ width: `${percentage}%` }}
                ></div>
              </div>
              <div className="w-8 text-sm text-gray-600">{count}</div>
            </div>
          );
        })}
      </div>
      
      {/* Recent Reviews */}
      {summary.recent_reviews.length > 0 && (
        <div>
          <h4 className="font-medium text-gray-900 mb-2">Recent Reviews</h4>
          {summary.recent_reviews.map((review, index) => (
            <div key={index} className="mb-3 pb-3 border-b border-gray-200 last:border-0">
              <div className="flex justify-between items-center mb-1">
                <div className="font-medium">{review.member_name}</div>
                <div className="text-sm text-gray-500">{new Date(review.created_at).toLocaleDateString()}</div>
              </div>
              <Rating value={review.rating} readOnly size="small" />
              {review.feedback && (
                <p className="text-sm text-gray-700 mt-1">{review.feedback}</p>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
}
